scripts/geoid_tool.py

```diff
@@ -0,0 +1,254 @@
+    args = docopt(__doc__)
+    in_dir = args.get('<reference_data>')
+    in_dir = abspath(in_dir)
+
+    epsg = int(args.get('-e'))
+
+    transform_command = args.get('-t')
+
+    debug = args.get('-d')
+
+    asp_dir = args.get('-a')
+    if asp_dir:
+        asp_dir = abspath(asp_dir)
+        if basename(asp_dir) != 'bin':
+            asp_dir = join(asp_dir, 'bin')
+    else:
+        asp_dir = abspath(join('ASP', 'bin'))
+
+    # setup our directory structure
+    geoid_dir = join(dirname(in_dir), 'geoid')
+    os.makedirs(geoid_dir, exist_ok= True)
+
+    # setup logging
+    log_dir = join(geoid_dir, 'logs')
+    os.makedirs(log_dir, exist_ok= True)
+    log_prefix = 'geoid-tool'
+    old_logs = glob(join(log_dir, f'{log_prefix}*.log'))
+    if old_logs:
+        # gets last run number from the old logs
+        vnum = max([int(basename(i).split('.')[0].split('-')[-1].replace('r','')) for i in sorted(old_logs)]) + 1
+    else:
+        # otherwise sets run number to 1
+        vnum = 1
+
+    logging.basicConfig(level=logging.INFO,
+                        format=f"(geoid-tool {__name__} %(levelname)s) %(message)s",
+                        # saves out to log file and outputs to command line.
+                        handlers=[logging.FileHandler(join(log_dir, f'{log_prefix}-r{vnum}.log')),
+                                  logging.StreamHandler(sys.stdout)]
+                        )
+    
+    log = logging.getLogger(__name__)
+
+    if debug:
+        log.setLevel(logging.DEBUG)  
+
+    # check for white spaces
+    if len([i for i in glob(join(in_dir, '*')) if ' ' in i]) > 0:
+        log.warning('White spaces found in file paths. Try and remove them?')
+        replace_white_spaces(in_dir)
+
+    # Check what kind of file was given by user
+    user_file_type = None
+    pc_ext = {'.las', '.laz', '.LAS', '.LAZ'}
+    raster_ext = {'.tif', '.tiff', '.TIF', '.TIFF'}
+    extension = in_dir.lower().split('.')[-1]
+    if f".{extension}" in pc_ext:
+        user_file_type = 'Point Cloud'
+    elif f".{extension}" in raster_ext:
+        user_file_type = 'Raster'
+    else:
+        user_file_type = None
+        raise Exception(f'User data needs to be in following format:{pc_ext} or {raster_ext}')
+    
+
+
+    # Make sure correct transformation selected, is their lidar in geoid, and they used the "to_ellipsoid" command mistakenly
+    epsg_object = CRS.from_epsg(epsg)
+    datum = epsg_object.datum.name
+    datum = code_datum(datum)
+    if 'WGS'==datum in datum and transform_command == 'to_geoid':
+        raise Exception(f'Your target is to get to ellipsoid based on input. You should use "to_ellipsoid".')
+    elif 'NAD' ==datum and transform_command == 'to_ellipsoid':
+        raise Exception(f'Your target is to get to geoid based on input. You should use "to_geoid".')
+    else:
+        pass
+
+    # run main functions
+    # Either Point Cloud or Raster...
+    if user_file_type == 'Point Cloud':
```

Comment:
I am a fan of keeping the `main` method concise and define methods to call.

How do you feel about creating two separate methods for each case and pass in the required arguments?

Reply:
Yeah I think I like that. I'll have to review next week with fresh eyes though!